src/v/redpanda/application.cc

```diff
@@ -1717,3 +1729,71 @@ void application::start_kafka(
       "Started Kafka API server listening at {}",
       config::node().kafka_api());
 }
+
+/**
+ * Feature table is generally updated via controller, but we need it to
+ * be initialized very early in startup so that other subsystems (including
+ * e.g. the controller raft group) may rely on up to date knowledge of which
+ * feature bits are enabled.
+ */
+void application::load_feature_table_snapshot() {
+    auto val_bytes_opt = storage.local().kvs().get(
+      storage::kvstore::key_space::controller,
+      features::feature_table_snapshot::kvstore_key());
+
+    if (!val_bytes_opt) {
+        // No snapshot?  Probably we are yet to join cluster.
+        return;
+    }
+
+    features::feature_table_snapshot snap;
+    try {
+        snap = serde::from_iobuf<features::feature_table_snapshot>(
+          std::move(*val_bytes_opt));
+    } catch (...) {
+        // Do not block redpanda from starting if there is something invalid
+        // here: the feature table should get replayed eventually via
+        // the controller.
+        vlog(
+          _log.error,
+          "Exception decoding feature table snapshot: {}",
+          std::current_exception());
+#ifndef NDEBUG
+        vassert(false, "Snapshot decode failed");
+#endif
+        return;
+    }
+
+    auto my_version = features::feature_table::get_latest_logical_version();
+    if (my_version < snap.version) {
+        vlog(
+          _log.error,
+          "Incompatible downgrade detected!  My version {}, feature table {} "
+          "indicates that all nodes in cluster were previously >= that version",
+          my_version,
+          snap.version);
+        // From this point, it is undefined to whether this process will be able
+        // to decode anything it sees on the network or on disk.
+#ifndef NDEBUG
+        vassert(my_version >= snap.version, "Incompatible downgrade detected");
+#endif
+    } else if (my_version > snap.version) {
```

Comment:
Also need to return early here?

Reply:
In the case of an incompatible downgrade, we will load the futuristic feature table snapshot, and ignore any feature states that refers to features we don't know about.  We will think the current cluster version is > our own binary's idea of the latest version, but that's an accurate reflection of the cluster state: we only get away from that by refusing to start -- refusing to apply the snapshot wouldn't improve matters.

If we didn't apply the snapshot, we would still proceed to replay the controller log and end up with the same end state via `feature_update_cmd` records that would increment the cluster version up beyond our local version, and hit the path in feature_table::apply_action where unknown features are logged+ignored.

Reply:
Ah good point about applying the snapshot here vs failing to start. I suppose that raises a good question -- should we fail to start here? Why only hit the assert in debug mode?

Reply:
I see, makes sense. Might be worth updating the language in this comment to make it clear that the fall-through is intentional, and that the permissive "downgrade" here is something we're knowingly sustaining even if we don't broadly support it. It's tempting and easy to miss both of these points (that we want to apply the snapshot, that we don't want to fail the downgrade) and change this codepath with early returns/crashes.